cbounter/ht_common.c

```diff
@@ -0,0 +1,963 @@
+    0,                         /*tp_compare*/
+    0,                         /*tp_repr*/
+    0,                         /*tp_as_number*/
+    0,                         /*tp_as_sequence*/
+    0,                         /*tp_as_mapping*/
+    0,                         /*tp_hash */
+    0,                         /*tp_call*/
+    0,                         /*tp_str*/
+    0,                         /*tp_getattro*/
+    0,                         /*tp_setattro*/
+    0,                         /*tp_as_buffer*/
+    #if PY_MAJOR_VERSION >= 3
+    Py_TPFLAGS_DEFAULT,
+    #else
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_ITER,
+      /* tp_flags: Py_TPFLAGS_HAVE_ITER tells python to
+         use tp_iter and tp_iternext fields. */
+    #endif
+    "Internal HTC iterator object.",           /* tp_doc */
+    0,  /* tp_traverse */
+    0,  /* tp_clear */
+    0,  /* tp_richcompare */
+    0,  /* tp_weaklistoffset */
+    HT_VARIANT(_ITER_iter),  /* tp_iter: __iter__() method */
+    HT_VARIANT(_ITER_iternext)  /* tp_iternext: next() method */
+};
+
+static inline HT_VARIANT(_ITER_TYPE) *
+HT_VARIANT(_make_iterator)(HT_TYPE *self, char result_type)
+{
+    HT_VARIANT(_ITER_TYPE) * iterator = PyObject_New(HT_VARIANT(_ITER_TYPE), &HT_VARIANT(_ITER_TYPE_Type));
+    if (!iterator)
+        return NULL;
+
+    if (!PyObject_Init((PyObject *)iterator, &HT_VARIANT(_ITER_TYPE_Type))) {
+        Py_DECREF(iterator);
+        return NULL;
+    }
+
+    iterator->i = 0;
+    iterator->hashtable = self;
+    iterator->result_type = result_type;
+    Py_INCREF(self);
+    return iterator;
+}
+
+static HT_VARIANT(_ITER_TYPE) *
+HT_VARIANT(_HT_iter_KV)(HT_TYPE *self)
+{
+    return HT_VARIANT(_make_iterator)(self, ITER_RESULT_KV_PAIRS);
+}
+
+static HT_VARIANT(_ITER_TYPE) *
+HT_VARIANT(_HT_iter_K)(HT_TYPE *self)
+{
+    return HT_VARIANT(_make_iterator)(self, ITER_RESULT_KEYS);
+}
+
+
+static PyMethodDef HT_VARIANT(_methods)[] = {
+    {"increment", (PyCFunction)HT_VARIANT(_increment), METH_VARARGS,
+     "Adds a string to the counter."
+    },
+    {"cardinality", (PyCFunction)HT_VARIANT(_cardinality), METH_NOARGS,
+     "Returns an estimate for the number of distinct items inserted into the counter. Does not work correctly when values are deleted!"
+    },
+    {"total", (PyCFunction)HT_VARIANT(_total), METH_NOARGS,
+     "Returns sum of all counts in the counter."
+    },
+    {"items", (PyCFunction)HT_VARIANT(_HT_iter_KV), METH_NOARGS,
+     "Iterates over all key-value pairs."
+    },
+    {"update", (PyCFunction)HT_VARIANT(_update), METH_VARARGS,
+     "Adds all pairs from another counter, or adds all items from an iterable."
+    },
+    {"quality", (PyCFunction)HT_VARIANT(_quality), METH_NOARGS,
+     "Returns the current estimated overflow rating of the structure, calculated as (cardinality / available buckets)."
+    },
+    {"histo", (PyCFunction)HT_VARIANT(_print_histo), METH_NOARGS,
+     "Iterates over all key-value pairs."
```

Comment:
Docstring needs updating. Maybe make `_histo()` since if it's a debug only method.